[Async Refactoring] Better handle non-refutable patterns

hamishknight opened the pull request:
Keep track of patterns that bind multiple vars and print them out when converting an async call. If the parameter being bound isn't referenced elsewhere, we'll print the pattern inline as e.g:

```
let ((x, y), z) = await foo()
```

Otherwise, if the parameter is referenced elsewhere in the block we'll print the pattern out of line, such as:

```
let (res, z) = await foo()
let (x, y) = res
```

In addition, make sure to print var bindings out of line if there's also a let binding, e.g:

```
let x = await foo()
var y = x
```

This ensures any mutations to `y` doesn't affect `x`. If there's only a single var binding, we'll print it inline.

Finally this PR ensures that refutable patterns and unclassified switch cases are left unhandled rather than being silently added to the success block.

rdar://77802560

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test macOS

bnbarham commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -5,6 +5,305 @@ enum E : Error { case e }
+  // MIXED-TUPLE-RESULT-NEXT:   let ((x, y), z) = res
+  // MIXED-TUPLE-RESULT-NEXT:   let ((x1, _), z1) = res
+  // MIXED-TUPLE-RESULT-NEXT:   print("a", x, y, z)
+  // MIXED-TUPLE-RESULT-NEXT:   print("b", x, z)
```
Shouldn't this be `x1, z1`?

Could we just avoid the second pattern, similar to how for single params we just keep the first naming?

bnbarham replied:
Ah, I assume the answer is "sort of" here. Ie. it would get pretty complicated and probably not worth it - the first pattern could have _ while the second binds all, etc, etc.

hamishknight replied:
> Shouldn't this be `x1, z1`?

Good catch! TIL apparently we introduce implicit vars into case stmt bodies, so we need to walk back up to the "parent" var in the pattern to handle this.

hamishknight replied:
> Ah, I assume the answer is "sort of" here. Ie. it would get pretty complicated and probably not worth it - the first pattern could have _ while the second binds all, etc, etc.

Yeah, I'm not really convinced the extra complexity will be worth it

bnbarham replied:
Heh, and now TIL! And yeah, I agree re. the complexity.

bnbarham commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -5,6 +5,305 @@ enum E : Error { case e }
+    print(str1, str2, int1, int2)
+  }
+  // MULTIPLE-TUPLE-PARAM2:      do {
+  // MULTIPLE-TUPLE-PARAM2-NEXT:   var ((str1, str2), (int1, int2)) = try await multipleTupleParam()
```
What are your thoughts on keeping `strs`/`ints` and then binding `str*` and `int*` separately when there is mixed var/let?

hamishknight replied:
We could do that, though I wonder if it makes sense to apply more generally, e.g if its `var (x, y) = ...` but only `x` is a `var` in the closure body, should we generate the following?

```
let (x, y) = ...
var x = x
```

bnbarham replied:
Hmmm... that does seem a little weird 😆. Let's just leave as is I think.

bnbarham replied:
Okay, so it looked weird because that would be a redecl 🤦.

We get a warning if a variable is mutable and not mutated, which is a bit of a pain with the case patterns, ie. if we make `case (let x, var y) = res` into `var (x, y) = res` there would then be a warning for `x`. Probably okay as an initial implementation?

hamishknight replied:
Yeah it would actually need to something more like this for us to handle it:

```
let (x1, y) = ...
var x = x1
```

Given this is a more general thing, I think I'll leave it for now, but I'll file a radar for us to handle it.

bnbarham replied:
And turns out it isn't actually a warning, I just had other warnings re. no reads/writes. So... all good 👍

bnbarham commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -5,6 +5,305 @@ enum E : Error { case e }
+  completion()
+}
+
+// It's a little unfortunate that the second 'strs' gets renamed to 'strs1'
+// given the first one gets replaced with an inline pattern, but I don't expect
+// this to be a common occurrance.
```
I assume this is just because we're currently adding all the success param names no matter what. You could probably pull the `ToPrint` construction out of `addAwaitCall` and then use it in `prepareNames` instead. Or merge the two/something else. Up to you though, probably fine as is anyway 🤷

hamishknight replied:
Fixed in latest update

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6563,24 +6601,143 @@ class AsyncConverter : private SourceEntityWalker {
+  ///
+  /// \param CE The call expr to convert.
+  /// \param Args The arguments of the call expr.
+  /// \param SuccessBlock The nodes present in the succes block following the
```
`succes`

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6563,24 +6601,143 @@ class AsyncConverter : private SourceEntityWalker {
+    if (Patterns.size() != 1)
+      return nullptr;
+
+    // We only do inline printing for patterns that bind multiple vars. Patterns
+    // that only bind a single var are printed by the main logic.
```
Would it be more or less complicated to handle them the same way, rather than have two code paths?

hamishknight replied:
My initial feeling was that it probably wouldn't be any simpler to merge the two code paths as there's still special casing that we do for the single var case such as coalescing them. That being said, if we also want to better handle the `var` case you mention as well as the `prepareNames` logic, it might be worth it.

bnbarham commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -5,6 +5,305 @@ enum E : Error { case e }
+// NAME-COLLISION-NEXT:   print(a, b1)
+// NAME-COLLISION-NEXT:   return
+// NAME-COLLISION-NEXT: }
+
```
Probably also worth adding a nested callback case.

bnbarham commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -0,0 +1,411 @@
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=STRING-TUPLE-TYPED-RESULT %s
+  stringTupleResult { res in
+    switch res {
+    case .success((let x, let y): (String, String)):
```
I hadn't seen this before so tried it myself - I get a compile error for this and a few others in this file. I'm a little confused as to how this test isn't failing since the *other* %refactor-check-compiles should fail to compile.

bnbarham replied:
Looks like the file being compiled is... empty. I'm guessing that happened in https://github.com/apple/swift/pull/37528?

bnbarham replied:
Yep, fixed here: https://github.com/apple/swift/pull/37755

hamishknight commented:
Updated to unify the code paths for single-var patterns and multi-var patterns a little (though a lot of logic is only really interested in single-var patterns), and also changed the behavior to make sure we always print `var` bindings out of line if there are also other `let` bindings present, as the `var` bindings may need separate storage for any mutations made.

I've left the handling of mixed let and var bindings in a single pattern for a future PR though.

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -5309,7 +5304,7 @@ struct CallbackClassifier {
-    ThenBlock->addAllBindings(CallbackConditions, DiagEngine);
+    ThenBlock->addAllBindings(CallbackConditions);
     if (DiagEngine.hadAnyError())
       return;
```
Ah just noticed I didn't remove this check, will do along with any review feedback.

bnbarham commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -5,6 +5,412 @@ enum E : Error { case e }
+  // FALLBACK-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=MIXED-BINDINGS %s
+  stringTupleParam { strs, err in
```
I wonder how often this actually comes up 😆. Very nice though!

hamishknight replied:
Yeah this exact case probably won't come up haha, but the subset of cases it covers might do, I just wanted to make sure they all interacted with each other okay :)

bnbarham commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -5,6 +5,412 @@ enum E : Error { case e }
+  // MIXED-TUPLE-RESULT-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=MIXED-TUPLE-RESULT2 %s
+  mixedTupleResult { res in
```
There's probably a few tests in here that are a little superfluous given others but... eh.

At some point I'd like to update the %refactor* tools to be similar to the swift-ide-test batch completion ones given the number we're adding 😆.

hamishknight replied:
Ah, I think what happened here is I ended up adding the `MIXED-TUPLE-RESULT` test case after this one and then didn't notice the overlap. I'll tweak this one so it's at least testing something slightly different.

bnbarham reviewed (approved):
Might be worth running these changes through the stress tester first given they're fairly... extensive. But otherwise, LGTM - very cool!

hamishknight commented on test/refactoring/ConvertAsync/convert_pattern.swift:
```diff
@@ -5,6 +5,412 @@ enum E : Error { case e }
+  // OUT-OF-LINE-CASE-NEXT: print(str1, str2, strs)
+
+  // RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=FALLBACK %s
+  stringTupleParam { strs, err in
```
Yeah that's exactly why, and I definitely agree we could do better here. I'll see if I can handle it along with cases like rdar://78564388. For now, I'll add a FIXME and add another fallback test case that's a little more reasonable :)

hamishknight committed:
Explicitly handle error case in fallback logic

I don't believe this case currently can come up,
but leave it explicitly unhandled for now so we
don't perform an erroneous transform if it ever
comes up in the future.

hamishknight committed:
Add a couple of getSemanticsProviding calls

hamishknight committed:
Don't handle refutable patterns in async transform

These aren't currently supported.

hamishknight committed:
Don't handle unclassified switch cases for now

Previously we would silently add these to the
success block. For now, let's leave them
unhandled.

hamishknight committed:
Adapt ScopedDeclCollector to track number of references

Use a DenseMap to track the number of references to
a given decl in a scope.

hamishknight committed:
Allow SemaAnnotator to handle patterns

Add the necessary walking hooks, and fix
ReferenceCollector to use it.

hamishknight committed:
[AST] Add Pattern::hasAnyMutableBindings

hamishknight committed:
Generalise addTupleOf

Allow it to iterate over an arbitrary container
type.

hamishknight committed:
Handle parent vars in async transform

If we're in a case stmt body, any DeclRefExprs
to vars bound by a pattern will actually be to an
implicit var decl that's declared for the body. We
therefore need to walk to its "parent" to get to
the var referenced by the pattern, which will have
the correct entries in the naming and placeholder
maps.

hamishknight committed:
Better handle non-refutable patterns in async transform

Keep track of patterns that bind multiple vars and
print them out when converting an async call. If
the parameter being bound isn't referenced elsewhere,
we'll print the pattern inline as e.g:

```
let ((x, y), z) = await foo()
```

Otherwise, if the parameter is referenced elsewhere
in the block we'll print the pattern out of line,
such as:

```
let (res, z) = await foo()
let (x, y) = res
```

In addition, make sure to print var bindings out
of line if there's also a let binding, e.g:

```
let x = await foo()
var y = x
```

This ensures any mutations to y doesn't affect x.
If there's only a single var binding, we'll print
it inline.

rdar://77802560

hamishknight committed:
Strip an unused DiagnosticEngine param

hamishknight commented:
@swift-ci please test

hamishknight commented:
TIL apparently you can do

@swift-ci please SourceKit stress test

hamishknight commented:
Hah, that uncovered another ASTWalker bug

bnbarham commented:
> Hah, that uncovered another ASTWalker bug

Woo! (sort of :P)

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
Stress tester failure unrelated
